test/pylib/ldap_server.py

```diff
@@ -0,0 +1,189 @@
+cn: role1
+uniqueMember: uid=jsmith,ou=People,dc=example,dc=com
+uniqueMember: uid=cassandra,ou=People,dc=example,dc=com
+""", """dn: cn=role2,dc=example,dc=com
+objectClass: groupOfUniqueNames
+cn: role2
+uniqueMember: uid=cassandra,ou=People,dc=example,dc=com
+""", """dn: cn=role3,dc=example,dc=com
+objectClass: groupOfUniqueNames
+cn: role3
+uniqueMember: uid=jdoe,ou=People,dc=example,dc=com
+""", ]
+
+
+def can_connect(address, family=socket.AF_INET):
+    s = socket.socket(family)
+    try:
+        s.connect(address)
+        return True
+    except OSError as e:
+        if 'AF_UNIX path too long' in str(e):
+            raise OSError(e.errno, f"{str(e)} ({address})") from None
+        else:
+            return False
+    except:
+        return False
+
+
+def try_something_backoff(something):
+    sleep_time = 0.05
+    while not something():
+        if sleep_time > 30:
+            return False
+        time.sleep(sleep_time)
+        sleep_time *= 2
+    return True
+
+
+def make_saslauthd_conf(ip, port, instance_path):
+    """Creates saslauthd.conf with appropriate contents under instance_path.  Returns the path to the new file."""
+    saslauthd_conf_path = os.path.join(instance_path, 'saslauthd.conf')
+    with open(saslauthd_conf_path, 'w') as f:
+        f.write(f'ldap_servers: ldap://{ip}:{port}\nldap_search_base: dc=example,dc=com')
+    return saslauthd_conf_path
+
+
+def start_ldap(host: Host, port: int, instance_root: Path, toxyproxy_byte_limit: int):
+    tp_port = 8474
+
+    tp_log_file = open(instance_root.parent / 'toxiproxy_server.log', 'x')
+    tp_server = subprocess.Popen(
+        ['toxiproxy-server', '-host', host, '-port', str(tp_port)],
+        stdout=tp_log_file,
+        stderr=subprocess.STDOUT
+    )
+
+    def can_connect_to_toxiproxy():
+        return can_connect((host, tp_port))
+
+    if not try_something_backoff(can_connect_to_toxiproxy):
+        raise Exception('Could not connect to toxiproxy')
+    
+    instance_path = instance_root / str(port)
+    slapd_pid_file = instance_path / 'slapd.pid'
+    saslauthd_socket_path = TemporaryDirectory()
+    os.makedirs(instance_path)
+    # This will always fail because it lacks the permissions to read the default slapd data
+    # folder but it does create the instance folder so we don't want to fail here.
+    try:
+        subprocess.check_output(['slaptest', '-f', TOP_SRC_DIR / LDAP_SERVER_CONFIGURATION_FILE, '-F', instance_path],
+                                stderr=subprocess.DEVNULL)
+    except:
+        pass
+    # Set up failure injection.
+    proxy_name = f'p{port}'
+    subprocess.check_output(
+        ['toxiproxy-cli', '-h', f'{host}:{tp_port}', 'c', proxy_name, '--listen', f'{host}:{port + 2}', '--upstream',
+         f'{host}:{port}'])
+    subprocess.check_output(
+        ['toxiproxy-cli', '-h', f'{host}:{tp_port}', 't', 'a', proxy_name, '-t', 'limit_data', '-n', 'limiter', '-a',
+         f'bytes={toxyproxy_byte_limit}'])
```

Comment:
@xtrey you are not capturing the stderr here. but you stated that you did at https://github.com/scylladb/scylladb/pull/23333#discussion_r2002815507 .

Reply:
I've added catching stdout on the server side. Do you think catching it here will be beneficial as well?

Reply:
toxiproxy-cli prints error to stderr. so, if we want to collect the error printed by it, we need to either redirect its stderr to stdout, like i did in #23333, or use something like:
```py
try:
  subprocess.check_output(['...'], stderr=subprocess.PIPE)
except subprocess.CalledProcessError as e
  logging.error("toxiproxy-cli failed: {}: {} {}", e.stdout, e.stderr)
```
but i'd prefer the former approach -- to redirect stderr to stdout, so we can just dump `e.stdout`.

Reply:
done